src/Symfony/Component/Messenger/Event/AbstractWorkerMessageEvent.php

```diff
@@ -36,4 +37,12 @@ public function getReceiverName(): string
     {
         return $this->receiverName;
     }
+
+    /**
+     * Add stamps to the envelope.
+     */
+    public function addStamps(StampInterface ...$stamps): void
+    {
+        $this->envelope = $this->envelope->with(...$stamps);
```

Comment:
This means all the events extending from AbstractWorkerMessageEvent can change the envelope. I think it's fine but the code handling events need to fetch the new envelope with getEnvelope before using the old envelope.
See https://github.com/symfony/symfony/blob/9bfa25869adab00162c246f4b76d65ca3b78e41a/src/Symfony/Component/Messenger/Worker.php#L103-L113

Reply:
The same applies to handling WorkerMessageHandledEvent and WorkerMessageFailedEvent

Reply:
This is a good point. Should I make the changes for this or should we open this up for a more general discussion?

Reply:
I've moved the `addStamps` method `WorkerMessageFailedEvent` and renamed it `addErrorDetailsStamp`. It now only takes a single `ErrorDetailsStamp` object, which allowed me to move the duplication checking logic from the event handler to this new method.

The only drawback is that I had to make the `$envelope` in `AbstractWorkerMessageEvent` `protected` instead of `private`.

Reply:
And... I've just had Timo re-add this method :). It's needed also on #36152. So basically, we have 2 use-cases where stamps need to be added to the envelope in an event listener. We've done the "conservative" thing to not give direct access to replace the entire envelope inside listeners (instead, only allowing stamps to be added), but this DOES mean that the Envelope is mutable, which means we will need to grab the Envelope in `Worker` as Tobion says.